api/v1alpha1/catalogentry_types.go

```diff
@@ -0,0 +1,108 @@
+	InternalAPIBindingEntryLabelKey = "internal.catalog.kcp.dev/entry"
+)
+
+// These are valid conditions of CatalogEntry.
+const (
+	// CatalogEntryValid is a condition for CatalogEntry that reflects the validity
+	// of the referenced APIExport.
+	CatalogEntryValid conditionsv1alpha1.ConditionType = "CatalogEntryValid"
+	// CatalogEntryInvalidReferenceReason is a reason for the CatalogEntryValid
+	// condition of APIBinding that the referenced CatalogEntry reference is invalid.
+	CatalogEntryInvalidReferenceReason = "CatalogEntryInvalidReference"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// CatalogEntry is the Schema for the catalogentries API
+type CatalogEntry struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   CatalogEntrySpec   `json:"spec,omitempty"`
+	Status CatalogEntryStatus `json:"status,omitempty"`
+}
+
+// CatalogEntrySpec defines the desired state of CatalogEntry
+type CatalogEntrySpec struct {
+	// references is a list of references to APIExports.
+	References []kcpv1alpha1.ExportReference `json:"references"`
+	// icon is the representation data that can be displayed in UI
+	// +optional
+	Icon Icon `json:"icon,omitempty"`
+	// description is a human-readable message to describe the information regarding
+	// the capabilities and features that the API provides
+	// +optional
+	Description string `json:"description,omitempty"`
+}
+
+// Icon is the graphical representation that can be displayed in UI component
+type Icon struct {
+	// data is the encoded icon data
+	Data []byte `json:"data"`
+	// mediaType is media type information for icon data
+	MediaType string `json:"mediaType"`
+}
+
+// CatalogEntryStatus defines the observed state of CatalogEntry
+type CatalogEntryStatus struct {
+	// permissionClaims is a list of permissions requested by the API provider(s)
+	// for this catalog entry.
+	// +optional
+	PermissionClaims []kcpv1alpha1.PermissionClaim `json:"PermissionClaims,omitempty"`
+	// apiResources is the list of APIs that are provided by this catalog entry.
+	// +optional
+	APIResources []metav1.GroupResource `json:"apiResources,omitempty"`
+	// conditions is a list of conditions that apply to the CatalogEntry.
+	//
+	// +optional
+	Conditions conditionsv1alpha1.Conditions `json:"conditions,omitempty"`
+}
+
+func (in *CatalogEntry) GetConditions() conditionsv1alpha1.Conditions {
+	return in.Status.Conditions
+}
+
+func (in *CatalogEntry) SetConditions(conditions conditionsv1alpha1.Conditions) {
+	in.Status.Conditions = conditions
+}
+
+//+kubebuilder:object:root=true
+
+// CatalogEntryList contains a list of CatalogEntry
+type CatalogEntryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []CatalogEntry `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&CatalogEntry{}, &CatalogEntryList{})
```

Comment:
Is it normal to put this into an API library?

Reply:
I think its fairly typical for Go APIs to export an `AddToScheme` function that adds all of the kinds in a particular group/version to the scheme. Not sure we need `controller-runtime`'s SchemeBuilder for that purpose though.

Reply:
This is generated code for the controller. I can punt it to the next PR when we add controller work.